sys/include/xtimer/implementation.h

```diff
@@ -65,7 +65,21 @@ static inline uint32_t _xtimer_lltimer_mask(uint32_t val)
  * @brief xtimer internal stuff
  * @internal
  */
+
 uint64_t _xtimer_now64(void);
+
+/**
+ * @brief Sets the timer to the appropriate timer_list or list_head.
+ *
+ * @note    The target to set the timer to has to be at least bigger then the
+ *          ticks needed to jump into the function and calculate '_xtimer_now()'.
+ *          So that 'now' did not pass the target.
```

Comment:
That is hard to guarantee I think. The problem is that the calling task may be de-scheduled for a few ticks and then you miss it anyway.

As it is a private function, is it not possible to update the API to pass it the `now` and `offset` value to handle it ?

Reply:
I really don't get this requirement. How can it be that for correct operation one needs to know the time it gets to get to a certain line? IMO this is very fragile.

Reply:
To not pass `now` and `offset`, another solution could be to just assume (and ensure) that `offset < UINT32_MAX - MAX_MARGIN` where `MAX_MARGIN` is the maximum time we allow to enter the function and read `now` again.

This way, you are sure that a timer loop can be detected if you took less than `MAX_MARGIN` to go to `now`.

A large enough solution could be to say it fits in an `int32_t` so uses `31bits`.
It is an easy thing to ensure over the code base.
With 1Mhz it would still allow working if you take 35 minutes between the first call and being in the function.

Reply:
This is no problem to guarantee. `XTIMER_BACKOFF` has to be set to as much ticks as it needs to reach the backing-off case. 

I allready started an test which helps to extract the required values.
See: https://github.com/RIOT-OS/RIOT/pull/8990/files

This requirement is based on the fact that the time is ticking while you try to set a timer. If you set your deadline too short you will miss it. Thats it.

Reply:
I will look at #8990 as a base.


When you start calling `_xtimer_set_absolute` you are still in your context, and you could be de-scheduled, for some time (maybe more than `XTIMER_BACKOFF`), so you do not know that the `target` is not elapsed, or not too close from your calling function.
Of course if you call the function with something that ends up being in the past, it should _shoot it directly, but you cannot do better as it was asked too late.

All the calling functions of `_xtimer_set_absolute` already have a `now` value (2 functions + 1 test).
By getting `now` and `offset` you remove the issues with "calling offset was too short" issue and handle this in the function directly.


However I see another issue also with the current implementation, it is that when `now` is read in the function, and `offset` is evaluated it is not with masked interrupts. So everything could also go wrong there too, be de-scheduled and get stuck with `xtimer_spin_until`.


The only hard limitation we cannot do anything about, is that between the first `now` and the `now` in the function, there should not be more than `UINT32_MAX / Frequency` ticks.

I wanted to say 2min23 for 32Mhz timers, but it looks like there are only sub MHz timers.
Even if there is this comment about cc2538 using a 16Mhz timer and code to handle them.

https://github.com/RIOT-OS/RIOT/blob/77dc923d7e96d8333a442d11e0e3e7d6563d3c12/sys/include/xtimer/tick_conversion.h#L41

``` bash
git grep XTIMER_HZ  . ':!sys'
boards/common/arduino-atmega/include/board_common.h:#define XTIMER_HZ                   (250000UL)
boards/frdm-kw41z/include/board.h:#define XTIMER_HZ                   (32768ul)
boards/hifive1/include/periph_conf.h:#define XTIMER_HZ                   (32768ul)
boards/ikea-tradfri/include/board.h:#define XTIMER_HZ           (250000UL)
boards/jiminy-mega256rfr2/include/board.h:#define XTIMER_HZ           (125000UL)
boards/mega-xplained/include/board.h:#define XTIMER_HZ                   (125000UL)
boards/mulle/include/board.h:#define XTIMER_HZ                   (32768ul)
boards/slstk3401a/include/board.h:#define XTIMER_HZ           (250000UL)
boards/slstk3402a/include/board.h:#define XTIMER_HZ           (250000UL)
boards/sltb001a/include/board.h:#define XTIMER_HZ           (250000UL)
boards/slwstk6000b/include/board.h:#define XTIMER_HZ           (250000UL)
boards/stk3600/include/board.h:#define XTIMER_HZ           (250000UL)
boards/stk3700/include/board.h:#define XTIMER_HZ           (250000UL)
boards/waspmote-pro/include/board.h:#define XTIMER_HZ                   (62500UL)
cpu/esp32/periph/timer.c:    CHECK_PARAM_RET (freq == XTIMER_HZ_BASE, -1);
cpu/esp32/periph/timer.c:    CHECK_PARAM_RET (freq == XTIMER_HZ_BASE, -1);
cpu/esp8266/periph/timer.c:    CHECK_PARAM_RET (freq == XTIMER_HZ_BASE, -1);
cpu/esp8266/periph/timer.c:    CHECK_PARAM_RET (freq == XTIMER_HZ_BASE, -1);
tests/bench_timers/Makefile:test-xtimer: CFLAGS+=-DTEST_XTIMER -DTIM_TEST_FREQ=XTIMER_HZ -DTIM_TEST_DEV=XTIMER_DEV
```


Looking at all this really needs a lot of concentration D:

Reply:
I could propose something with my ideas and see if it would help.

Reply:
Did you see #7630 regarding passing a reference time as argument to _xtimer_set_absolute?
It needs a rebase but it is related to this discussion I think.

Reply:
I did not, I just jumped in the xtimer issue to provide a test, but it means that I had a similar reasoning. I will provide a minimal test to show an error.

Reply:
@gebart I think your fix still misses something, from what I get, the function should be calling `xtimer_now` from within a masked interrupt context, so that the time between this `now` and the `periph_timer_set` can be guaranteed to be fixed as you cannot be de-scheduled.

If not the comparison with `XTIMER_BACKOFF` does not guarantee anything.

Reply:
I will try to show these with tests.